### PR TITLE
[PERF] Enhance local allgather for multiple rounds

### PR DESCRIPTION
**Pull request by ZYHowell**
This is for https://github.com/alpa-projects/alpa/issues/350. The algorithm is in comments of the issue.

### CODE REVIEW

File: alpa/pipeline_parallel/decentralized_distributed_runtime.py [hunk 1]
```diff
@@ -732,6 +732,7 @@ def _compile_resharding_task(self, src_mesh_idx: int, dst_mesh_idx: int,
         # add recv task for each worker
         for w, task_uuid in resharding_task.recv_worker_task_ids.items():
             output_uuids = recv_buf_uuids[w]
+            # TODO(yonghao)
```
Comment (merrymercy) [inline alpa/pipeline_parallel/decentralized_distributed_runtime.py:735]: what's this?

**Review by merrymercy — CHANGES_REQUESTED** [review]
This part is getting very complicated and hard to follow. Let us try to organize it better.

1. Move the description of the algorithm from github posts to the code comments. Add a more high-level algorithm description at the beginning of the file.
2. Define important data structure at the very beginning of the file. Move utility functions and class (e.g.,` _suffix_prod`) to the end of the file.
3. Create unit tests in `test_cross_mesh_resharding.py`. In each test case,
    1. explicitly create a `ReshardingTask`
    2. compile a resharding strategy
    3. assert the optimizations are applied (e.g. load balance, local all-gather)
    4. run it and test the correctness
   In the future, a new optimization should come with new test cases, so we know what cases can be optimized by the new optimization.

**Comment by zhisbug** [comment]
I probably need a while to review this PR :) @ZYHowell

**Comment by ZYHowell** [comment]
>     3. Create unit tests in `test_cross_mesh_resharding.py`. In each test case,
>        
>        1. explicitly create a `ReshardingTask`
>        2. compile a resharding strategy
>        3. assert the optimizations are applied (e.g. load balance, local all-gather)
>        4. run it and test the correctness
> 
> 
> In the future, a new optimization should come with new test cases, so we know what cases can be optimized by the new optimization.

See https://github.com/alpa-projects/alpa/pull/360.

**Comment by merrymercy** [comment]
Let us try to get this reviewed or merged soon

**Comment by merrymercy** [comment]
Let us try to get this reviewed or merged soon. @zhisbug @ZYHowell

**Review by zhisbug — APPROVED** [review]
